File: src/vault/migrate.ts
import { readFile, rename } from 'fs/promises';
import { existsSync } from 'fs';
import { homedir, hostname, userInfo } from 'os';
import { join } from 'path';
import { createDecipheriv, scryptSync } from 'crypto';
import { CliError } from '../utils/errors';
import type { Config } from '../types/config';
import type { StoredCredentials } from '../types/tokens';

function configDir(): string {
  return join(process.env.HOME || homedir(), '.config', 'agentio');
}

export function legacyPaths(): { configPath: string; tokensPath: string } {
  const dir = configDir();
  return {
    configPath: join(dir, 'config.json'),
    tokensPath: join(dir, 'tokens.enc'),
  };
}

export async function detectLegacy(): Promise<{ hasConfig: boolean; hasTokens: boolean }> {
  const { configPath, tokensPath } = legacyPaths();
  return {
    hasConfig: existsSync(configPath),
    hasTokens: existsSync(tokensPath),
  };
}

function legacyMachineKey(): Buffer {
  const machineId = `${hostname()}-${userInfo().username}-agentio-v1`;
  return scryptSync(machineId, 'agentio-salt', 32);
}

function tryDecryptLegacyTokens(raw: string): StoredCredentials | null {
  try {
    const { iv, tag, data } = JSON.parse(raw);
    const key = legacyMachineKey();
    const decipher = createDecipheriv('aes-256-gcm', key, Buffer.from(iv, 'hex'));
    decipher.setAuthTag(Buffer.from(tag, 'hex'));
    const plaintext = Buffer.concat([
      decipher.update(Buffer.from(data, 'hex')),
      decipher.final(),
    ]);
    return JSON.parse(plaintext.toString('utf-8'));
  } catch {
    return null;
  }
}

export interface MigrateResult {
  config: Config;
  credentials: StoredCredentials;
  tokensRecovered: boolean;
}

/**
 * Read legacy config + tokens without side effects. Safe to call multiple
 * times. Call archiveLegacy() after the new vault has been successfully
 * written to preserve the originals as .bak files.
 */
export async function readLegacy(): Promise<MigrateResult> {
  const { configPath, tokensPath } = legacyPaths();

  if (!existsSync(configPath)) {
    throw new CliError(
      'NOT_FOUND',
      'No legacy config.json to migrate',
      'Run: agentio vault init for a fresh install'
    );
  }

  let config: Config;
  try {
    const raw = await readFile(configPath, 'utf-8');
    config = JSON.parse(raw) as Config;
  } catch (err) {
    throw new CliError(
      'CONFIG_ERROR',
      `Failed to read legacy config.json: ${(err as Error).message}`,
      'Fix or remove the file, then run setup again'
    );
  }

  let credentials: StoredCredentials = {};
  let tokensRecovered = false;
  if (existsSync(tokensPath)) {
    const raw = await readFile(tokensPath, 'utf-8');
    const recovered = tryDecryptLegacyTokens(raw);
    if (recovered) {
      credentials = recovered;
      tokensRecovered = true;
    }
  }

  return { config, credentials, tokensRecovered };
}

/** Rename legacy files to .bak. Call only after the new vault is safely written. */
export async function archiveLegacy(): Promise<void> {
  const { configPath, tokensPath } = legacyPaths();
  if (existsSync(configPath)) {
    await rename(configPath, configPath + '.bak');
  }
  if (existsSync(tokensPath)) {
    await rename(tokensPath, tokensPath + '.bak');
  }
}

/** Back-compat wrapper: read then archive in one step. */
export async function migrateLegacy(): Promise<MigrateResult> {
  const result = await readLegacy();
  await archiveLegacy();
  return result;
}

File: go/internal/vault/migrate.go
package vault

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"encoding/hex"
	"encoding/json"
	"os"
	"os/user"
	"path/filepath"

	"github.com/plosson/agentio/go/internal/clierr"
	"golang.org/x/crypto/scrypt"
)

// Legacy files from before the single vault. Init reads them once, writes the
// new vault, then renames the originals to .bak.

// LegacyPaths are the pre-vault config.json and tokens.enc.
func LegacyPaths() (configPath, tokensPath string) { return legacyPaths() }

func legacyPaths() (configPath, tokensPath string) {
	dir := ConfigDir()
	return filepath.Join(dir, "config.json"), filepath.Join(dir, "tokens.enc")
}

func DetectLegacy() (hasConfig, hasTokens bool) {
	cfg, tok := legacyPaths()
	_, err1 := os.Stat(cfg)
	_, err2 := os.Stat(tok)
	return err1 == nil, err2 == nil
}

type MigrateResult struct {
	Config          Config
	Credentials     map[string]map[string]map[string]any
	TokensRecovered bool
}

func ReadLegacy() (*MigrateResult, error) {
	cfgPath, tokPath := legacyPaths()
	if _, err := os.Stat(cfgPath); err != nil {
		return nil, clierr.New(clierr.NotFound, "No legacy config.json to migrate", "Run: agentio vault init for a fresh install")
	}
	raw, err := os.ReadFile(cfgPath)
	if err != nil {
		return nil, clierr.New(clierr.ConfigError, "Failed to read legacy config.json: "+err.Error(), "Fix or remove the file, then run setup again")
	}
	var cfg Config
	if err := json.Unmarshal(raw, &cfg); err != nil {
		return nil, clierr.New(clierr.ConfigError, "Failed to read legacy config.json: "+err.Error(), "Fix or remove the file, then run setup again")
	}
	if cfg.Profiles == nil {
		cfg.Profiles = map[string][]ProfileValue{}
	}
	result := &MigrateResult{Config: cfg, Credentials: map[string]map[string]map[string]any{}}
	if b, err := os.ReadFile(tokPath); err == nil {
		if creds := tryDecryptLegacy(string(b)); creds != nil {
			result.Credentials = creds
			result.TokensRecovered = true
		}
	}
	return result, nil
}

func ArchiveLegacy() error {
	cfg, tok := legacyPaths()
	for _, path := range []string{cfg, tok} {
		if _, err := os.Stat(path); err != nil {
			continue
		}
		if err := AssertTestWritable(path, "legacy config"); err != nil {
			return err
		}
		if err := os.Rename(path, path+".bak"); err != nil {
			return err
		}
	}
	return nil
}

func tryDecryptLegacy(raw string) map[string]map[string]map[string]any {
	var blob struct {
		IV   string `json:"iv"`
		Tag  string `json:"tag"`
		Data string `json:"data"`
	}
	if err := json.Unmarshal([]byte(raw), &blob); err != nil {
		return nil
	}
	iv, err1 := hex.DecodeString(blob.IV)
	tag, err2 := hex.DecodeString(blob.Tag)
	data, err3 := hex.DecodeString(blob.Data)
	if err1 != nil || err2 != nil || err3 != nil {
		return nil
	}
	key, err := legacyKey()
	if err != nil {
		return nil
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil
	}
	aead, err := cipher.NewGCMWithNonceSize(block, len(iv))
	if err != nil {
		return nil
	}
	sealed := append(append([]byte{}, data...), tag...)
	plain, err := aead.Open(nil, iv, sealed, nil)
	if err != nil {
		return nil
	}
	dec := json.NewDecoder(bytes.NewReader(plain))
	dec.UseNumber()
	var creds map[string]map[string]map[string]any
	if err := dec.Decode(&creds); err != nil {
		return nil
	}
	return creds
}

func legacyKey() ([]byte, error) {
	host, err := os.Hostname()
	if err != nil {
		return nil, err
	}
	u, err := user.Current()
	if err != nil {
		return nil, err
	}
	machine := host + "-" + u.Username + "-agentio-v1"
	return scrypt.Key([]byte(machine), []byte("agentio-salt"), scryptN, scryptR, scryptP, keyLen)
}

// RemoveLegacyBackups deletes the config.json.bak and tokens.enc.bak a
// migration left, as Bun's vault reset does. A failed delete is ignored.
func RemoveLegacyBackups() error {
	configPath, tokensPath := legacyPaths()
	for _, p := range []string{configPath + ".bak", tokensPath + ".bak"} {
		if _, err := os.Stat(p); err != nil {
			continue
		}
		if err := AssertTestWritable(p, "legacy backup"); err != nil {
			return err
		}
		_ = os.Remove(p)
	}
	return nil
}
